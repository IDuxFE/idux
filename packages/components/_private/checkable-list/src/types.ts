/**
 * @license
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://github.com/IDuxFE/idux/blob/main/LICENSE
 */

import type { VirtualScrollToFn } from '@idux/cdk/scroll'
import type { ExtractInnerPropTypes, ExtractPublicPropTypes } from '@idux/cdk/utils'
import type { DefineComponent, HTMLAttributes } from 'vue'

import { IxPropTypes, type VKey } from '@idux/cdk/utils'

// eslint-disable-next-line @typescript-eslint/no-explicit-any
export interface CheckableListData extends Record<VKey, any> {
  key?: VKey
  label?: string
  disabled?: boolean
  additional?: {
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    class?: any
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    style?: any
    [key: string]: unknown
  }
}

export interface CheckableListScroll {
  height?: string | number
  fullHeight?: boolean
}

export const checkableListProps = {
  dataSource: IxPropTypes.array<CheckableListData>(),
  checkable: IxPropTypes.bool.def(true),
  removable: IxPropTypes.bool.def(false),
  checked: IxPropTypes.func<(item: CheckableListData) => boolean>(),
  disabled: IxPropTypes.func<(item: CheckableListData) => boolean>(),
  getRowKey: IxPropTypes.func<(item: CheckableListData) => VKey>(),
  virtual: IxPropTypes.bool,
  scroll: IxPropTypes.object<CheckableListScroll>(),
  onCheckChange: IxPropTypes.emit<(item: CheckableListData, checked: boolean) => void>(),
  onRemove: IxPropTypes.emit<(item: CheckableListData) => void>(),
  onScroll: IxPropTypes.emit<(evt: Event) => void>(),
  onScrolledChange: IxPropTypes.emit<(startIndex: number, endIndex: number, visibleData: unknown[]) => void>(),
  onScrolledBottom: IxPropTypes.emit<() => void>(),
}

export const checkableListItemProps = {
  checked: IxPropTypes.bool.def(false),
  checkable: IxPropTypes.bool.def(true),
  removable: IxPropTypes.bool.def(false),
  disabled: IxPropTypes.bool.def(false),
  label: IxPropTypes.string,
  value: IxPropTypes.oneOfType([String, Number, Symbol]).isRequired,
  onCheckChange: IxPropTypes.emit<(checked: boolean) => void>(),
  onRemove: IxPropTypes.emit<() => void>(),
}

export interface CheckableListApi {
  scrollTo: VirtualScrollToFn
}

export type CheckableListProps = ExtractInnerPropTypes<typeof checkableListProps>
export type CheckableListItemProps = ExtractInnerPropTypes<typeof checkableListItemProps>
export type CheckableListPublicProps = ExtractPublicPropTypes<typeof checkableListProps>
export type CheckableListComponent = DefineComponent<
  Omit<HTMLAttributes, keyof CheckableListPublicProps> & CheckableListPublicProps,
  CheckableListApi
>
export type CheckableListInstance = InstanceType<DefineComponent<CheckableListProps, CheckableListApi>>
