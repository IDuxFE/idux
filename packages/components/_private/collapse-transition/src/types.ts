/**
 * @license
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://github.com/IDuxFE/idux/blob/main/LICENSE
 */

import type { ExtractInnerPropTypes, ExtractPublicPropTypes } from '@idux/cdk/utils'
import type { DefineComponent, HTMLAttributes } from 'vue'

import { IxPropTypes } from '@idux/cdk/utils'

export type CollapseTransitionMode = 'height' | 'width'

export const collapseTransitionProps = {
  appear: IxPropTypes.bool.def(false),
  name: IxPropTypes.string,
  mode: IxPropTypes.oneOf<CollapseTransitionMode>(['height', 'width']).def('height'),
  onAfterEnter: IxPropTypes.emit(),
  onAfterLeave: IxPropTypes.emit(),
}

export type CollapseTransitionProps = ExtractInnerPropTypes<typeof collapseTransitionProps>
export type CollapseTransitionPublicProps = ExtractPublicPropTypes<typeof collapseTransitionProps>
export type CollapseTransitionComponent = DefineComponent<
  Omit<HTMLAttributes, keyof CollapseTransitionPublicProps> & CollapseTransitionPublicProps
>
export type CollapseTransitionInstance = InstanceType<DefineComponent<CollapseTransitionProps>>
