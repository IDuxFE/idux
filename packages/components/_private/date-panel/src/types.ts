/**
 * @license
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://github.com/IDuxFE/idux/blob/main/LICENSE
 */

import type { ExtractInnerPropTypes, ExtractPublicPropTypes, MaybeArray } from '@idux/cdk/utils'
import type { DefineComponent, HTMLAttributes, PropType } from 'vue'

export const datePanelProps = {
  cellTooltip: Function as PropType<(cell: { value: Date; disabled: boolean }) => string | void>,
  disabledDate: Function as PropType<(date: Date) => boolean>,
  type: {
    type: String as PropType<DatePanelType>,
    default: 'date',
  },
  value: [Date, Array] as PropType<Date | (Date | undefined)[]>,
  activeDate: Date,
  visible: {
    type: Boolean,
    default: undefined,
  },
  isSelecting: {
    type: Boolean,
    default: undefined,
  },

  onCellClick: [Function, Array] as PropType<MaybeArray<(date: Date) => void>>,
  onCellMouseenter: [Function, Array] as PropType<MaybeArray<(date: Date) => void>>,
  'onUpdate:activeDate': [Function, Array] as PropType<MaybeArray<(date: Date) => void>>,
}

export type DatePanelProps = ExtractInnerPropTypes<typeof datePanelProps>
export type DatePanelPublicProps = ExtractPublicPropTypes<typeof datePanelProps>
export type DatePanelComponent = DefineComponent<
  Omit<HTMLAttributes, keyof DatePanelPublicProps> & DatePanelPublicProps
>
export type DatePanelInstance = InstanceType<DefineComponent<DatePanelProps>>

export type DatePanelType = 'date' | 'week' | 'month' | 'quarter' | 'year'

// private
export const panelRowProps = {
  rowIndex: {
    type: Number,
    required: true,
  },
}

export const panelCellProps = {
  rowIndex: {
    type: Number,
    required: true,
  },
  cellIndex: {
    type: Number,
    required: true,
  },
}
