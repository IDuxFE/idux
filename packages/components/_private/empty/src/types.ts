/**
 * @license
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://github.com/IDuxFE/idux/blob/main/LICENSE
 */

import type { EmptyProps as IxEmptyProps } from '@idux/components/empty'

export interface EmptyProps {
  empty?: string | IxEmptyProps
}
