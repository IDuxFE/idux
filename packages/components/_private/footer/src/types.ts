/**
 * @license
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://github.com/IDuxFE/idux/blob/main/LICENSE
 */

import type { ExtractInnerPropTypes, ExtractPublicPropTypes, VKey } from '@idux/cdk/utils'
import type { ButtonProps } from '@idux/components/button'
import type { DefineComponent, HTMLAttributes, PropType, VNode } from 'vue'

export interface FooterButtonProps extends ButtonProps {
  key?: VKey
  text?: string | VNode
  onClick?: (evt: Event) => void
}

export const footerProps = {
  cancel: Function as PropType<(evt?: Event | unknown) => Promise<void> | void>,
  cancelButton: Object as PropType<ButtonProps>,
  cancelLoading: Boolean,
  cancelText: String,
  cancelVisible: {
    type: Boolean,
    default: true,
  },
  footer: [Boolean, Array, Object] as PropType<boolean | FooterButtonProps[] | VNode>,
  ok: Function as PropType<(evt?: Event | unknown) => Promise<void> | void>,
  okButton: Object as PropType<ButtonProps>,
  okLoading: Boolean,
  okText: String,
} as const

export type FooterProps = ExtractInnerPropTypes<typeof footerProps>
export type FooterPublicProps = ExtractPublicPropTypes<typeof footerProps>
export type FooterComponent = DefineComponent<Omit<HTMLAttributes, keyof FooterPublicProps> & FooterPublicProps>
export type FooterInstance = InstanceType<DefineComponent<FooterProps>>
