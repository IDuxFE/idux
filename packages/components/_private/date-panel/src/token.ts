/**
 * @license
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://github.com/IDuxFE/idux/blob/main/LICENSE
 */

/* eslint-disable @typescript-eslint/no-explicit-any */

import type { ActiveDateContext } from './composables/useActiveDate'
import type { ActiveTypeContext } from './composables/useActiveType'
import type { MaxIndexContext } from './composables/useMaxIndex'
import type { DatePanelProps } from './types'
import type { DateConfig } from '@idux/components/config'
import type { DatePickerLocale } from '@idux/components/i18n'
import type { ComputedRef, InjectionKey, Slots } from 'vue'

export interface DatePanelContext extends ActiveTypeContext, ActiveDateContext, MaxIndexContext {
  props: DatePanelProps
  slots: Slots
  locale: ComputedRef<DatePickerLocale>
  mergedPrefixCls: ComputedRef<string>
  dateConfig: DateConfig
}

export const datePanelToken: InjectionKey<DatePanelContext> = Symbol('datePanelToken')
