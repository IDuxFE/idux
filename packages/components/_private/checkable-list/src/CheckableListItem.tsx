/**
 * @license
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://github.com/IDuxFE/idux/blob/main/LICENSE
 */

import { computed, defineComponent, inject, normalizeClass } from 'vue'

import { callEmit } from '@idux/cdk/utils'
import { IxCheckbox } from '@idux/components/checkbox'
import { IxIcon } from '@idux/components/icon'

import { checkableListContext } from './token'
import { checkableListItemProps } from './types'

export default defineComponent({
  props: checkableListItemProps,
  setup(props, { slots }) {
    const { mergedPrefixCls } = inject(checkableListContext)!

    const onCheckChange = (value: number | boolean | string) => {
      callEmit(props.onCheckChange, !!value)
    }
    const onRemove = () => {
      callEmit(props.onRemove)
    }

    const renderLabel = (prefixCls: string) => {
      const { checked, value, disabled, checkable } = props

      if (checkable) {
        return (
          <IxCheckbox
            class={`${prefixCls}-checkbox`}
            v-slots={slots}
            label={props.label}
            value={value}
            checked={checked}
            disabled={disabled}
            onChange={onCheckChange}
          />
        )
      }

      return <label class={`${prefixCls}-label`}>{slots.default?.() ?? props.label}</label>
    }

    const classes = computed(() => {
      const prefixCls = `${mergedPrefixCls.value}-item`

      return normalizeClass({
        [prefixCls]: true,
        [`${prefixCls}-disabled`]: props.disabled,
      })
    })

    return () => {
      const prefixCls = `${mergedPrefixCls.value}-item`

      return (
        <li class={classes.value}>
          {renderLabel(prefixCls)}
          {props.removable && !props.disabled && (
            <IxIcon class={`${prefixCls}-close-icon`} name="close" onClick={onRemove} />
          )}
        </li>
      )
    }
  },
})
