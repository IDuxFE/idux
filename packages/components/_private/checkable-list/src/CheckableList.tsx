/**
 * @license
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://github.com/IDuxFE/idux/blob/main/LICENSE
 */

import { computed, defineComponent, normalizeClass, provide, ref } from 'vue'

import { CdkVirtualScroll, type VirtualItemRenderFn, type VirtualScrollInstance } from '@idux/cdk/scroll'
import { callEmit } from '@idux/cdk/utils'
import { useGlobalConfig } from '@idux/components/config'

import CheckableListItem from './CheckableListItem'
import { checkableListContext } from './token'
import { type CheckableListApi, type CheckableListData, checkableListProps } from './types'

export default defineComponent({
  name: 'ÉµCheckableList',
  props: checkableListProps,
  setup(props, { slots, expose }) {
    const common = useGlobalConfig('common')
    const mergedPrefixCls = computed(() => `${common.prefixCls}-checkable-list`)
    const labelKey = computed(() => props.labelKey ?? 'label')
    const virtualScrollRef = ref<VirtualScrollInstance>()

    provide(checkableListContext, {
      mergedPrefixCls,
    })

    const checkableListApi: CheckableListApi = {
      scrollTo: (...params) => virtualScrollRef.value?.scrollTo(...params),
    }

    expose(checkableListApi)

    const getKey = (item: CheckableListData) => (props.getKey?.(item) ?? item.key)!

    const handleScroll = (evt: Event) => {
      callEmit(props.onScroll, evt)
    }
    const handleScrolledBottom = () => {
      callEmit(props.onScrolledBottom)
    }
    const handleScrolledChange = (startIndex: number, endIndex: number, visibleData: unknown[]) => {
      callEmit(props.onScrolledChange, startIndex, endIndex, visibleData)
    }

    const renderListItem: VirtualItemRenderFn<CheckableListData> = ({ item, index }) => {
      const key = getKey(item)
      const onCheckChange = (checked: boolean) => {
        callEmit(props.onCheckChange, item, checked)
      }
      const onRemove = () => {
        callEmit(props.onRemove, item)
      }

      const customAdditional = props.customAdditional ? props.customAdditional({ data: item, index }) : undefined

      return (
        <CheckableListItem
          key={key}
          value={key!}
          label={item[labelKey.value] as string}
          checked={!!props.checked?.(item)}
          disabled={!!props.disabled?.(item)}
          checkable={props.checkable}
          removable={props.removable}
          v-slots={{ default: slots.label && (() => slots.label?.(item)) }}
          onCheckChange={onCheckChange}
          onRemove={onRemove}
          {...item.additional}
          {...customAdditional}
        />
      )
    }

    const renderBody = () => {
      const { dataSource, virtual, scroll } = props
      const data = dataSource ?? []

      if (data.length <= 0) {
        return
      }

      if (virtual && scroll) {
        const { height, fullHeight } = scroll
        return (
          <CdkVirtualScroll
            ref={virtualScrollRef}
            dataSource={data}
            fullHeight={fullHeight}
            getKey={getKey}
            height={height as number}
            itemHeight={32}
            itemRender={renderListItem}
            virtual
            onScroll={handleScroll}
            onScrolledBottom={handleScrolledBottom}
            onScrolledChange={handleScrolledChange}
          />
        )
      }

      return (
        <ul class={`${mergedPrefixCls.value}-inner`} onScroll={handleScroll}>
          {data.map((item, index) => renderListItem({ item, index }))}
        </ul>
      )
    }

    const classes = computed(() => {
      const prefixCls = mergedPrefixCls.value

      return normalizeClass({
        [prefixCls]: true,
        [`${prefixCls}-virtual`]: !!props.virtual,
      })
    })

    return () => {
      return <div class={classes.value}>{renderBody()}</div>
    }
  },
})
