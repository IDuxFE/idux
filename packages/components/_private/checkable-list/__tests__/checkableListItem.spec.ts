import { MountingOptions, mount } from '@vue/test-utils'
import { computed, h } from 'vue'

import { renderWork } from '@tests'

import { IxCheckbox } from '@idux/components/checkbox'

import CheckableListItem from '../src/CheckableListItem'
import { checkableListContext } from '../src/token'
import { CheckableListItemProps } from '../src/types'

const mountGlobalOpts = {
  provide: {
    [checkableListContext as symbol]: {
      mergedPrefixCls: computed(() => 'ix-checkable-list'),
    },
  },
}

describe('CheckableListItem', () => {
  const CheckableListItemMount = (options?: MountingOptions<Partial<CheckableListItemProps>>) =>
    mount(CheckableListItem, { ...(options as MountingOptions<CheckableListItemProps>), global: mountGlobalOpts })

  renderWork<CheckableListItemProps>(CheckableListItem, {
    props: {
      value: '1',
      label: 'Option-1',
      checkable: true,
      removable: true,
      disabled: false,
    },
    global: mountGlobalOpts,
  })

  test('checkable work', async () => {
    const wrapper = CheckableListItemMount({
      props: {
        value: '1',
        label: 'Option-1',
        checked: true,
        checkable: true,
        removable: true,
        disabled: false,
      },
    })

    const checkboxComp = wrapper.findComponent(IxCheckbox)
    expect(checkboxComp.exists()).toBeTruthy()
    expect(checkboxComp.vm.checked).toBe(true)

    await wrapper.setProps({ checkable: false })
    expect(wrapper.findComponent(IxCheckbox).exists()).toBeFalsy()
  })

  test('removable work', async () => {
    const onRemove = jest.fn()
    const wrapper = CheckableListItemMount({
      props: {
        value: '1',
        label: 'Option-1',
        checkable: true,
        removable: true,
        disabled: false,
        onRemove,
      },
    })

    const removeTrigger = wrapper.find('.ix-checkable-list-item-close-icon')
    expect(removeTrigger.exists()).toBeTruthy()
    await removeTrigger.trigger('click')
    expect(onRemove).toBeCalled()

    await wrapper.setProps({ removable: false })
    expect(wrapper.find('.ix-checkable-list-item-close-icon').exists()).toBeFalsy()
  })

  test('disabled work', async () => {
    const onCheckChange = jest.fn()
    const wrapper = CheckableListItemMount({
      props: {
        value: '1',
        label: 'Option-1',
        checkable: true,
        removable: true,
        disabled: true,
        onCheckChange,
      },
    })

    expect(wrapper.classes()).toContain('ix-checkable-list-item-disabled')
    expect(wrapper.find('.ix-checkable-list-item-close-icon').exists()).toBeFalsy()
    await wrapper.findComponent(IxCheckbox).find('input').setValue(true)
    expect(onCheckChange).not.toBeCalled()

    await wrapper.findComponent(IxCheckbox).find('input').setValue(false)

    await wrapper.setProps({ disabled: false })
    expect(wrapper.classes()).not.toContain('ix-checkable-list-item-disabled')
    expect(wrapper.find('.ix-checkable-list-item-close-icon').exists()).toBeTruthy()
    await wrapper.findComponent(IxCheckbox).find('input').setValue(true)
    expect(onCheckChange).toBeCalledWith(true)
  })

  test('slot work', async () => {
    const wrapper = CheckableListItemMount({
      props: {
        value: '1',
        label: 'Option-1',
        checkable: true,
        removable: true,
        disabled: true,
      },
      slots: {
        default: () => h('span', { class: 'custom-label-slot' }),
      },
    })

    expect(wrapper.find('.custom-label-slot').exists()).toBeTruthy()
  })
})
