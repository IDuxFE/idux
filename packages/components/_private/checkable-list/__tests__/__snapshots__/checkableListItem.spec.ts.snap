// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CheckableListItem render work 1`] = `
"<li class=\\"ix-checkable-list-item\\"><label class=\\"ix-checkbox ix-checkable-list-item-checkbox\\" role=\\"checkbox\\" aria-checked=\\"false\\" aria-disabled=\\"false\\"><span class=\\"ix-checkbox-input\\"><input type=\\"checkbox\\" class=\\"ix-checkbox-input-inner\\" aria-hidden=\\"true\\" value=\\"1\\"><span class=\\"ix-checkbox-input-box\\"></span></span><span class=\\"ix-checkbox-label\\">Option-1</span>
    <!---->
  </label><i class=\\"ix-icon ix-icon-close ix-checkable-list-item-close-icon\\" role=\\"img\\" aria-label=\\"close\\"></i></li>"
`;
